// @apollo/gateway/dist/utilities/graphql.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseSelections = exports.astFromType = exports.allNodesAreOfSameKind = exports.getResponseName = void 0;
const graphql_1 = require("graphql");
const assert_1 = require("./assert");
function getResponseName(node) {
    return node.alias ? node.alias.value : node.name.value;
}
exports.getResponseName = getResponseName;
function allNodesAreOfSameKind(firstNode, remainingNodes) {
    return !remainingNodes.some(node => node.kind !== firstNode.kind);
}
exports.allNodesAreOfSameKind = allNodesAreOfSameKind;
function astFromType(type) {
    if (graphql_1.isListType(type)) {
        return { kind: graphql_1.Kind.LIST_TYPE, type: astFromType(type.ofType) };
    }
    else if (graphql_1.isNonNullType(type)) {
        return { kind: graphql_1.Kind.NON_NULL_TYPE, type: astFromType(type.ofType) };
    }
    else {
        return {
            kind: graphql_1.Kind.NAMED_TYPE,
            name: { kind: graphql_1.Kind.NAME, value: type.name },
        };
    }
}
exports.astFromType = astFromType;
function parseSelections(source) {
    const parsed = graphql_1.parse(`{${source}}`);
    assert_1.assert(parsed.definitions.length === 1, `Invalid FieldSet provided: '${source}'. FieldSets may not contain operations within them.`);
    return parsed.definitions[0].selectionSet
        .selections;
}
exports.parseSelections = parseSelections;
