// @apollo/query-planner/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueryPlanner = exports.buildOperationContext = exports.prettyFormatQueryPlan = exports.astSerializer = exports.queryPlanSerializer = void 0;
var snapshotSerializers_1 = require("./snapshotSerializers");
Object.defineProperty(exports, "queryPlanSerializer", { enumerable: true, get: function () { return snapshotSerializers_1.queryPlanSerializer; } });
Object.defineProperty(exports, "astSerializer", { enumerable: true, get: function () { return snapshotSerializers_1.astSerializer; } });
var prettyFormatQueryPlan_1 = require("./prettyFormatQueryPlan");
Object.defineProperty(exports, "prettyFormatQueryPlan", { enumerable: true, get: function () { return prettyFormatQueryPlan_1.prettyFormatQueryPlan; } });
__exportStar(require("./QueryPlan"), exports);
__exportStar(require("./composedSchema"), exports);
const buildQueryPlan_1 = require("./buildQueryPlan");
Object.defineProperty(exports, "buildOperationContext", { enumerable: true, get: function () { return buildQueryPlan_1.buildOperationContext; } });
class QueryPlanner {
    constructor(schema) {
        this.schema = schema;
    }
    buildQueryPlan(operationContext, options) {
        return buildQueryPlan_1.buildQueryPlan({ ...operationContext, schema: this.schema }, options);
    }
}
exports.QueryPlanner = QueryPlanner;
