// @apollo/federation/dist/composition/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.findDirectivesOnNode = exports.isStringValueNode = exports.compositionHasErrors = exports.defaultRootOperationNameLookup = exports.normalizeTypeDefs = exports.compositionRules = void 0;
__exportStar(require("./compose"), exports);
__exportStar(require("./composeAndValidate"), exports);
__exportStar(require("./types"), exports);
var rules_1 = require("./rules");
Object.defineProperty(exports, "compositionRules", { enumerable: true, get: function () { return rules_1.compositionRules; } });
var normalize_1 = require("./normalize");
Object.defineProperty(exports, "normalizeTypeDefs", { enumerable: true, get: function () { return normalize_1.normalizeTypeDefs; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "defaultRootOperationNameLookup", { enumerable: true, get: function () { return utils_1.defaultRootOperationNameLookup; } });
Object.defineProperty(exports, "compositionHasErrors", { enumerable: true, get: function () { return utils_1.compositionHasErrors; } });
Object.defineProperty(exports, "isStringValueNode", { enumerable: true, get: function () { return utils_1.isStringValueNode; } });
Object.defineProperty(exports, "findDirectivesOnNode", { enumerable: true, get: function () { return utils_1.findDirectivesOnNode; } });
