// @apollo/gateway/dist/__generated__/graphqlTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageLevel = exports.FetchErrorCode = exports.ErrorCode = void 0;
var ErrorCode;
(function (ErrorCode) {
    ErrorCode["InvalidBody"] = "INVALID_BODY";
    ErrorCode["UnexpectedResponse"] = "UNEXPECTED_RESPONSE";
    ErrorCode["ConnectionFailed"] = "CONNECTION_FAILED";
    ErrorCode["Timeout"] = "TIMEOUT";
    ErrorCode["Other"] = "OTHER";
})(ErrorCode = exports.ErrorCode || (exports.ErrorCode = {}));
var FetchErrorCode;
(function (FetchErrorCode) {
    FetchErrorCode["AuthenticationFailed"] = "AUTHENTICATION_FAILED";
    FetchErrorCode["AccessDenied"] = "ACCESS_DENIED";
    FetchErrorCode["UnknownRef"] = "UNKNOWN_REF";
    FetchErrorCode["RetryLater"] = "RETRY_LATER";
})(FetchErrorCode = exports.FetchErrorCode || (exports.FetchErrorCode = {}));
var MessageLevel;
(function (MessageLevel) {
    MessageLevel["Error"] = "ERROR";
    MessageLevel["Warn"] = "WARN";
    MessageLevel["Info"] = "INFO";
})(MessageLevel = exports.MessageLevel || (exports.MessageLevel = {}));
